docs/proposals/cross-tenant-query-federation.md

```diff
@@ -0,0 +1,99 @@
+---
+title: "Cross-Tenant Query Federation"
+linkTitle: "Cross-Tenant Query Federation"
+weight: 1
+slug: "cross-tenant-query-federation"
+---
+
+- Author: [Christian Simon](https://github.com/simonswine)
+- Date: October 2020
+- Status: Proposed
+
+## Overview
+
+This document aims to describe how to implement the ability to allow queries to cover data from more than a single Cortex tenant.
+
+## Reasoning
+
+Adopting a tenancy model within an organization with each tenant representing a department comes with the disadvantage that it will prevent queries from spanning multiple departments. This proposal tries to overcome those limitations.
+
+## Alternatives considered
+
+In theory PromQL API clients could be aggregating/correlating query results from multiple tenants. For example Grafana could be used with multiple data sources and a cross tenant query could be achieved through using [Transformations][grafana_transformation].
+
+[grafana_transformation]: https://grafana.com/docs/grafana/latest/panels/transformations/
+
+## Challenges
+
+### Aggregate data without overlaps
+
+#### Challenge
+
+The series in different tenants might have exactly the same labels and hence potentially collide which each other.
+
+#### Proposal
+
+In order to be able to always identify the tenant correctly, queries using multiple tenants should inject a tenant label and its value containing the tenant ID into the results. An potentially existing label with the same name should be stored in a label with the prefix `original_`. This also should be done recursively until no more conflicts arise.
```

Comment:
> An potentially existing label with the same name should be stored in a label with the prefix `original_`.

Is this complexity required? In Prometheus, label names starting with `__` are considered internal / reserved. We could just use the label name `__tenant__` and override any existing one by design (but we expect it doesn't exist).

Reply:
You can see the use case for the configurable label/a discussion here: https://github.com/cortexproject/cortex/pull/3364#discussion_r508762843. In the case of making the label configurable the complexity is required.

The use case for a configurable tenant label is something like team, organization, product, etc... could be an obvious tenant separation and much clearer to the end user as to where that metric is coming from. For a MVP only allowing `__tenant__` would be acceptable to me though, I would just like it to be configurable in the future.

Reply:
I'm in favor of restricting the label to be `__tenant__`. I think the design around configuring the injected label and recursively prefixing should be move to a **Future Work** section.

Reply:
:+1: on moving to future work.

Reply:
I have amended the proposal accordingly. Thanks for your comments.